Strip binaries more aggressively

janvorli opened the pull request:
We are currently stripping only debug symbols, but there are also other
symbols that can be stripped without affecting debuggability or anything
else. We were stripping the binaries in corefx that way in the past,
but some time ago we have unified the stripping between the former
three repos and started to strip just the debugging symbols.

This change strips all unneeded symbols, which results in significant
reduction of the binaries sizes. The table below shows the reduction 
for x64 Linux:

| File | Before | After | Delta |
|-|-:|-:|-:|
| libSystem.IO.Compression.Native.so | 921680 | 805280 | 116400 |
| libSystem.IO.Ports.Native.so | 22552 | 18528 | 4024 |
| libSystem.Native.so | 95256 | 71784 | 23472 |
| libSystem.Net.Security.Native.so | 19112 | 14488 | 4624 |
| libSystem.Security.Cryptography.Native.OpenSsl.so | 163376 | 120952 | 42424 | 
| libclrgc.so | 854528 | 695976 | 158552 |
| libclrjit.so | 3323160 | 3093656 | 229504 |
| libcoreclr.so | 8732272 | 7027024 | 1705248 |
| libmscordaccore.so | 3174184 | 2568936 | 605248 |
| libmscordbi.so | 2389368 | 1828536 | 560832 |
| dotnet | 97368 | 72272 | 25096 |
| apphost | 103512 | 77352 | 26160 |
| libhostfxr.so | 416304 | 330184 | 86120 |
| libhostpolicy.so | 382720 | 293336 | 89384 |
| **Total** | **20695392** | **17018304** | **3677088** |

janvorli committed:
Strip binaries more aggressively

We are currently stripping only debug symbols, but there are also other
symbols that can be stripped without affecting debuggability or anything
else. We were stripping the binaries in corefx that way in the past,
but some time ago we have unified the stripping between the former
three repos and started to strip just the debugging symbols.

This change strips all unneeded symbols, which results in significant
reduction of the binaries sizes. For example, libcoreclr.so went down
from 8732272 to 7027024 or libSystem.Security.Cryptography.Native.OpenSsl.so
went down from 163376 to 120952.

janvorli commented:
Contributes to fixing #39598

jkotas reviewed (approved):
Nice!

janvorli commented:
FreeBSD has a problem with this level of stripping symbols - it cannot find the g_dacTable symbol. I'm looking into it.

janvorli committed:
Fix FreeBSD

janvorli commented:
Hmm, the fix for FreeBSD breaks OSX, since there are no dynamic symbols on OSX. I'll need to fix it after my vacation.

ViktorHofer commented:
@janvorli do you think this will still be ready for 5.0 which closes soon? Presumably this change would need to go through tactics for 5.0.

janvorli commented:
@ViktorHofer yes, I am working on it right now.

ViktorHofer commented:
Awesome, thanks!

janvorli committed:
Fix FreeBSD and OSX

We use nm command to extract g_dacTable symbol value from libcoreclr
for the DAC purposes. On OSX, there are no dynamic symbols, on
FreeBSD, the stripping strips all regular symbols and leaves
dynamic ones only. So the nm needs to get a -D argument for
non OSX.

janvorli commented:
@am11 would you be able to verify that this change is correct for SunOS?

am11 commented:
Sure, I am running in fork CI https://github.com/am11/runtime/runs/1058972514 for illumos x64.

am11 commented:
CoreCLR crossbuild for illumos-x64 has succeeded (symbols stripping steps went normally). Thank you for the heads up @janvorli!

janvorli commented:
@am11 thanks a lot for verifying it!

janvorli commented:
/backport to release/5.0-rc2

CoffeeFlux commented:
Do we want to do something similar in mono, at least for the mobile SDKs? @marek-safar

marek-safar commented:
@akoeplinger worked on that for 5.0 wasm where it makes the biggest impact. We should do the same for mobile even though we run native linker on the output for most of the configurations.
